_posts/2018/2018-02-18-avoid-writing-sql-when-using-activerelation.md

```diff
@@ -0,0 +1,76 @@
+---
+title: 'Avoid Writing SQL When Using ActiveRelation'
+description: 'Why write SQL when you can write Ruby?'
+layout: article
+category: ruby
+image:
+  base: '2018/avoid-writing-sql-when-using-activerelation'
+  alt: 'Coding SQL'
+  source: 'https://unsplash.com/photos/fPkvU7RDmCo'
+  credit: 'Caspar Rubin'
+---
+
+ActiveRelation, which powers the searching and querying engine of ActiveRecord, is a powerful and flexible tool.
+
+
+## Instead of...
+
+...writing literal SQL strings with direct string interpolation inside an ActiveRelation `#where` method.
+
+```ruby
+Person.where("name = #{ params[:name] } AND hidden_at IS NULL")
+```
+
+
+## Or...
+
+...writing literal SQL strings and using the ‘array style’ to safely interpolate user input.
+
+```ruby
+Person.where('name = ? AND hidden_at IS NULL', params[:name])
+```
+
+
+## Use...
+
+...the ‘hash style’ syntax.
+
+```ruby
+Person.where(name: params[:name], hidden_at: nil)
+```
+
+
+## Why?
+
+The first two versions, where you manually write SQL, were the _only_ way before ActiveRelation was merged into Rails (in version 3.0). However, the “hash style” gives you more flexibility and safety.
+
+The first example is **very** dangerous. It uses direct interpolation of the string, in this case from user passed parameters. **Do not use this style**. It opens you up to SQL injection attacks: where bad people from the Internet can try and run destructive/exciting statements against your database.
+
+The second example, the ‘array style’, does sanitize passed data, so is a better way to write “string style” `#where` methods. However you're still left to write SQL yourself, when `ActiveRelation` is more than happy to generate perfect SQL while you keep writing Ruby.
```

Comment:
*write SQL yourself. Why go there when `ActiveRelation` is more than happy to generate perfect SQL while you focus on writing perfect (could be funny)/excellent/wonderful/friendly (or maybe none of the above) Ruby?

(Or something like that?)